src/modules/storage/BlockStorage.ts

```diff
@@ -0,0 +1,128 @@
+/*******************************************************************************
+
+    The class that insert and read the block into the database.
+
+    Now only block header are stored.
+    Transactions and other data will be managed by different classes.
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+import * as sqlite from 'sqlite3';
+
+/**
+ * The class that insert and read the block into the database.
+ */
+class BlockStorage
+{
+    /**
+     *  The instance of sqlite3
+     */
+    private db: sqlite.Database;
+
+    /**
+     * Constructor
+     * @param filename Valid values are filenames,
+     * ":memory:" for an anonymous in-memory database and
+     * an empty string for an anonymous disk-based database
+     * @param callback : If provided, this function will be called when
+     * the database was opened successfully or when an error occurred.
+     * The first argument is an error object.
+     */
+    constructor (filename : string, callback : any = null)
+    {
+        this.db = new sqlite.Database(filename,
+            sqlite.OPEN_CREATE | sqlite.OPEN_READWRITE |
+            sqlite.OPEN_SHAREDCACHE);
+        this.db.configure("busyTimeout", 1000);
+        this.createTable(callback);
+    }
+
+    /**
+     * Creates a table. table name is `block`.
+     * @param callback If provided, this function will be called when
+     * the database was finished successfully or when an error occurred.
+     * The first argument is an error object.
+     */
+    public createTable (callback : any = null)
+    {
+        var sql =
+        `CREATE TABLE IF NOT EXISTS block
+        (
+            height INTEGER PRIMARY KEY,
+            prev_block TEXT,
+            merkle_root TEXT,
+            signature TEXT
+        )`;
+        this.db.run(sql, (err : any) =>
+        {
+            if (callback != null)
+                callback(err);
+        });
+    }
+
+    /**
+     * Puts a block to database
+     * @param data a block data
+     * @param callback If provided, this function will be called when
+     * the database was finished successfully or when an error occurred.
+     * The first argument is an error object.
+     */
+    public put (data : any, callback : any = null)
+    {
+        var sql =
+        `INSERT INTO block
+            (height, prev_block, merkle_root, signature)
+        VALUES
+            (?, ?, ?, ?)
+        ON CONFLICT(height) DO UPDATE SET
```

Comment:
Is there a scenario where it is possible to update the data if it already exists?

Reply:
This shouldn't have happened.
However, if this happens, it is to avoid the error caused by it.
If this can happen, data recovery by polling may be added in the future.

https://www.sqlite.org/lang_UPSERT.html